examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1

```diff
@@ -0,0 +1,180 @@
+   $vmList = ${cloudEventData}.virtualMachines
+   $dryRun = ${cloudEventData}.dryRun
+
+   # Check to ensure at least one snapshot policy is defined
+   if($maxSnapshotCount -eq "" -and $maxSnapshotSize -eq "" -and $maxSnapshotDays -eq "") {
+      throw "No snapshot retentation policy were defined"
+   }
+
+   # Check to ensure we have a list of VMs
+   if($vmList -eq "" -or $vmList.getType().BaseType.Name -ne "Array") {
+      throw "Invalid or missing input found for VM list"
+   }
+
+   # Retrieve list of VMs to run snapshot management against
+   try {
+      $vms = Get-Vm -NoRecursion -Name $vmList
+   } catch {
+      throw "Failed to retrieve VM list"
+   }
+
+   foreach ($vm in $vms) {
+      $vmName = ${vm}.name
+
+      Write-Host "$(Get-Date) - Checking VM: $vmName "
+      try {
+         $snapshots = $vm | Get-Snapshot
+      } catch {
+         throw "Failed to retrieve snapshots from VM $vmName "
+      }
+
+      if($maxSnapshotSize -ne "") {
+         foreach ($snapshot in $snapshots) {
+            $snapshotName = ${snapshot}.name
+
+            # Get snapshot size (GB) and check against max size
+            if($snapshot.sizeGB -gt $maxSnapshotSize) {
+               Write-Host "$(Get-Date) - `tSnapshot $snapshotName consumes $(${snapshot}.SizeGB) GB and exceeds maximum size (${maxSnapshotSize} GB)"
+
+               if($dryRun -eq $false) {
+                  Write-Host "$(Get-Date) - `tSnapshot removal started for $snapshotName"
+
+                  try {
+                     $snapshot | Remove-Snapshot -RunAsync -Confirm:$false
+                  } catch {
+                     throw "Failed to remove snapshot $snapshotName from VM $vmName with error $($Error[0])"
+                  }
+               }
+            }
+         }
+      }
+
+      if($maxSnapshotDays -ne "") {
+         $currentDate = Get-Date
+
+         foreach ($snapshot in $snapshots) {
+            $snapshotName = ${snapshot}.name
+
+            # Get snapshot creation date & check against current date
+            $snapshotDate = $snapshot.Created
+            $snapshotAge = ($currentDate - $snapshotDate).Days
+
+            if($snapshotAge -gt $maxSnapshotDays) {
+               Write-Host "$(Get-Date) - `tSnapshot $snapshotName is $snapshotAge days old and exceeds maximum number of days (${maxSnapshotDays})"
+
+               if($dryRun -eq $false) {
+                  Write-Host "$(Get-Date) - `tSnapshot removal started for $snapshotName"
+
+                  try {
+                     $snapshot | Remove-Snapshot -RunAsync -Confirm:$false
+                  } catch {
+                     throw "Failed to remove snapshot $snapshotName from VM $vmName with error $($Error[0])"
+                  }
+               }
+            }
+         }
+      }
+
+      if($maxSnapshotCount -ne "") {
+         if($snapshots.count -gt $maxSnapshotCount) {
+
+            # Get the older snapshot
+            $snapshot = $snapshots | Select-Object -Last 1
```

Comment:
Shouldn't you remove all snaps (i.e. diff) between max and current.count? Also, "-Last" is sorted by oldest here? For me "-Last" reads like the last (most recent) one :)

Reply:
hm ... logic is indeed invalid right now. It shouldn't remove all but basically the "diff" as you mentioned. 

```
$diff = $maxSnapshotCount - $snapshots.count
$snapsToRemove = $snapshots | Select-Object -Last $diff
```

This way, you're removing the n-number of oldest but this also brings up a good question on whether users want to delete from newest or oldest within snapshot tree ...

Reply:
I think the logic to handle count might be tricky. For sake of simplicity, I'm going to remove this use case